src/bloom.rs

```diff
@@ -0,0 +1,104 @@
+use bitvec::prelude::*;
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+
+/// Bloom implements bloom filter functionalities over
+/// a bit-slice of data.
+pub struct Bloom<'a> {
+    filter: &'a BitSlice<Lsb0, u8>,
+    k: u32,
+}
+
+impl<'a> Bloom<'a> {
+    /// Create a bloom filter from a byte slice
+    pub fn new(buf: &'a [u8]) -> Self {
+        let filter = &buf[..buf.len() - 4];
+        let k = (&buf[buf.len() - 4..]).get_u32();
+        Self {
+            filter: &BitSlice::from_slice(filter).unwrap(),
+            k,
+        }
+    }
+
+    /// Get bloom filter bits per key from entries count and FPR
+    pub fn bloom_bits_per_key(entries: usize, false_positive_rate: f64) -> usize {
+        let size =
+            -1.0 * (entries as f64) * false_positive_rate.ln() / (0.69314718056 as f64).powi(2);
+        let locs = 0.69314718056 * size / (entries as f64);
+        locs as usize
+    }
+
+    /// Build bloom filter from key hashes
+    pub fn build_from_key_hashes(keys: &[u32], bits_per_key: usize) -> Bytes {
```

Comment:
can we make it return `Self` directly?
and rename `new` into `from_bytes`?

Reply:
`Bloom` do not own any data by design. So we could not return `Self`.

Reply:
The bloom filter is stored inside `TableIndex`, which is a protobuf struct. Therefore, we we are accessing table index, we could only get a byte sequence representing a bloom filter. By using current scheme, we could construct a bloom without any copy.